### PR TITLE
Fix Issue 21456 - std.format does not accept enum member with string base type as template parameter

### PR DESCRIPTION
**Pull request by berni44**
Don't know if it wouldn't be better to have to return type just being `auto`...

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
> Don't know if it wouldn't be better to have to return type just being auto

if that works then yes please

File: std/format/package.d [hunk 1]
```diff
@@ -1341,12 +1341,10 @@ if (isSomeChar!Char)
...
 
 /// ditto
-typeof(fmt) format(alias fmt, Args...)(Args args)
-if (isSomeString!(typeof(fmt)))
+auto format(alias fmt, Args...)(Args args)
+if (is(StringTypeOf!(typeof(fmt))))
 {
     import std.array : appender;
-    import std.range.primitives : ElementEncodingType;
-    import std.traits : Unqual;
```
Comment (RazvanN7) [inline std/format/package.d:1349]: Why are these imports made public?

Comment (berni44) [reply std/format/package.d:1352]: Sorry, I needed that for stating the return type explicitly and forgot to undo this, when using `auto`.

**Review by RazvanN7 — APPROVED** [review]

**Comment by andralex** [comment]
This is distinctly the wrong kind of generality.

I am very sorry we missed the initial support for enum types derived from strings. No we don't need to support them. Whoever has those enums may as well call `representation` and call it a day.

This pull should be undone and all support for enum strings should be deprecated.

**Comment by berni44** [comment]
> This is distinctly the wrong kind of generality.
> 
> I am very sorry we missed the initial support for enum types derived from strings. No we don't need to support them. Whoever has those enums may as well call `representation` and call it a day.
> 
> This pull should be undone and all support for enum strings should be deprecated.

For me it is OK to revert this PR - I just tried to care about a bug and marking this bug as invalid would achieve my goals too.

Anyway, I would like to understand what is wrong with enum types derived from strings? If I would have known this, I would have probably marked the issue immediately as invalid and gave a reason there. Maybe you can point me to a place, where this is explained? (Or explain it yourself in brief?)

**Comment by nordlow** [comment]
> This is distinctly the wrong kind of generality.
> 
> I am very sorry we missed the initial support for enum types derived from strings. No we don't need to support them. Whoever has those enums may as well call `representation` and call it a day.
> 
> This pull should be undone and all support for enum strings should be deprecated.

One great motive for ditching enum strings is that it avoids duplicated template instances.

For instance

```d
void format(alias fmt)() { pragma(msg, fmt); }
```

```d
unittest {
    format!("%s")();
    enum Fmt : string { s = "%s" }
    format!(Fmt.s)();
}
```
prints

```
%s
%s
```
meaning format is instantiated twice, completely unnecessary. This creates template bloat that slows down the compiler.

**Comment by Geod24** [comment]
I am wondering, why is `fmt` an `alias` and not a simple `string` ? The only reason I can think of is to support `wstring` and `dstring`, but I wouldn't find that compelling, as `fmt` has to be CT-known, and the main reasons to have non UTF8 string types is either interacting with other APIs or micro optimizing for memory usage.

Having `fmt` be a `string` would have avoided the need for special support here. From a quick look, it also seems that reverting this will prevent types with `alias this` from working. Not that I think it's a good idea to support it (or enum format string), but if we do, it shouldn't require special code to do so.

**Comment by berni44** [comment]
> I am wondering, why is `fmt` an `alias` and not a simple `string`?

Don't know. It was introduced in #5288 and not changed since. At least no unittest complains, when I replace it with `string fmt` and remove the constraint.

**Comment by Geod24** [comment]
CC @ntrel

**Comment by ntrel** [comment]
> why is fmt an alias and not a simple string ?

No real reason. I think I just used alias whilst working on the change as it's more flexible, then never thought to tighten it to string before submitting.

**Comment by schveiguy** [comment]
> Having `fmt` be a `string` would have avoided the need for special support here.

This is probably the right answer. It should work with anything that currently works, except `wstring` and `dstring` formats. For those, you can probably use a converter, or just generate `string` in the first place.

A good first step might be reverting this, and seeing if replacing `alias` with `string` breaks anything.

**Comment by schveiguy** [comment]
Note that we still would need to deprecate if we did move to accepting a string directly.

**Comment by nordlow** [comment]
Are there more format specificer being `alias` that should be `string` in the helper functions for format``?

**Pull request merged**